### ISSUE TITLE
openssh@8.5p1: ssh command broken

### ISSUE THREAD

**Issue by jsonbruce** [Homebrew/homebrew-core#72448]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```
HOMEBREW_VERSION: 3.0.4-12-g13508ee
ORIGIN: https://github.com/Homebrew/brew
HEAD: 13508eebaa18d09b1eeaeda88d6cb252c32d9f18
Last commit: 18 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: bf39da6c1356d50b09e206f8c8ac06da2b9a07ba
Core tap last commit: 15 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: dodeca-core 64-bit kabylake
Clang: 12.0 build 1200
Git: 2.30.1 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.2.2-x86_64
CLT: 12.4.0.0.1.1610135815
Xcode: 12.4
```

### 

- [X] The `brew doctor` above contains no "Warning" lines.

### What were you trying to do (and why)?

```
brew upgrade openssh
ssh root@remote-host
```

### What happened (include all command output)?

```
➜  brew upgrade openssh
==> Upgrading 1 outdated package:                                                                                                                      
openssh 8.4p1_2 -> 8.5p1                                                                                                                               
==> Upgrading openssh 8.4p1_2 -> 8.5p1                                                                                                                 
==> Downloading https://homebrew.bintray.com/bottles/openssh-8.5p1.big_sur.bottle.tar.gz                                                               
==> Downloading from https://d29vzk4ow07wi7.cloudfront.net/072d30cb24b4b9f9fb87096202cb007129a3299d0954cf06216b0b14a40f7d78?response-content-dispositio
######################################################################## 100.0%                                                                        
==> Pouring openssh-8.5p1.big_sur.bottle.tar.gz                                                                                                        
🍺  /usr/local/Cellar/openssh/8.5p1: 47 files, 5.9MB                                                                                                   
Removing: /usr/local/Cellar/openssh/8.4p1_2... (35 files, 4.9MB)                                                                                       
Removing: /usr/local/Cellar/openssh/8.4p1_2.reinstall... (47 files, 5.5MB)                                                                             
Removing: /Users/max/Library/Caches/Homebrew/openssh--8.4p1_2.big_sur.bottle.tar.gz... (1.9MB)                                                         
Removing: /Users/max/Library/Caches/Homebrew/openssh--8.4p1.tar.gz... (1.7MB)   

➜ ssh root@remote-host
dyld: Library not loaded: /usr/local/opt/ldns/lib/libldns.3.dylib
  Referenced from: /usr/local/bin/ssh
  Reason: image not found
[1]    77000 abort      ssh root@remote-host
```

### What did you expect to happen?

```
ssh to remote-host successfully.
```

### Step-by-step reproduction instructions (by running `brew` commands)

```
brew upgrade openssh
ssh root@remote-host
```

**Comment by jsonbruce** [comment]
I've gone back to the previous version `openssh@8.4p1`, which has no issues.

**Comment by carlocab** [comment]
Can't reproduce:
```
❯ brew install openssh
==> Downloading https://homebrew.bintray.com/bottles/ldns-1.7.1_3.big_sur.bottle.tar.gz
[snip]
❯ ssh root@remote-host
ssh: Could not resolve hostname remote-host: nodename nor servname provided, or not known
```

Try
```
brew update
brew reinstall openssh
```

If that doesn't work:
```
brew update
brew reinstall $(brew deps openssh) openssh
```

**Comment by Bo98** [comment]
There's a weird discrepency in `ldns` dylib versions. If you install the `ldns` Catalina bottle, you will get `libldns.0.dylib` but on Big Sur you will get `libldns.3.dylib`.

If you upgraded macOS versions, then you still likely have the older `ldns` build. `brew reinstall ldns` will work.

We probably should revision bump `ldns` to sort out the discrepency.

**Comment by mitchblank** [comment]
Yeah I hit the same problem here tonight.  A `brew reinstall ldns` was all that was needed to fix locally.  Didn't dig into what version the dylib has in the various bottles though.

**Comment by northway** [comment]
> Yeah I hit the same problem here tonight. A `brew reinstall ldns` was all that was needed to fix locally. Didn't dig into what version the dylib has in the various bottles though.

Confirmed, it works!

**Comment by metayan** [comment]
Same issue on Catalina.
```
brew reinstall ldns
```
fixed it.

**Comment by carlocab** [comment]
Looks like this is now resolved. We can reopen this if it turns out that's not the case.

**Comment by metayan** [comment]
Wouldn't say it's resolved. I had to `brew reinstall ldns` manually, even though I just did an `update/upgrade`. At least on Catalina.

**Comment by carlocab** [comment]
Ok, let me clarify. Anything here that isn't a duplicate of another open issue (e.g. https://github.com/Homebrew/brew/issues/10127) is resolved. Would you still say that's inaccurate?

**Comment by metayan** [comment]
Not sure what you mean, @carlocab . I use Catalina, didn't do an OS upgrade, did `brew update`, `brew upgrade`, but still `ssh` failed. I would call a resolution something that fixes the issue with an `update/upgrade`, without having to search here for what to do. Maybe a version bump on `ldns`? I'm not the expert.
`ssh` just suddenly failing can give plenty of people heart issues.
Just sayin'.

**Comment by carlocab** [comment]
I guess the difficulty here is that your issue looks to be something else entirely, and not the issue that this one was first opened about. Your issue needs to be tracked elsewhere and not tracked in a related but largely orthogonal one.

**Comment by metayan** [comment]
Same problem and same solution as reported in this issue.
Only difference being Catalina and not BS, which might explain the `ldns` version discrepancy below. (Apparently 3 for BS.)
```
ssh
dyld: Library not loaded: /usr/local/opt/ldns/lib/libldns.0.dylib
  Referenced from: /usr/local/bin/ssh
  Reason: image not found
Abort trap: 6
```
I don't know why you mentioned https://github.com/Homebrew/brew/issues/10127 in https://github.com/Homebrew/homebrew-core/issues/72448#issuecomment-792300944 . As far as I can tell it has nothing to do with this issue.
Plus, I don't have an issue anymore, but whoever will do an `update/upgrade` on Catalina and find `ssh` failing will probably expand their children's vocabulary. ;)
Peace and Love.

**Comment by carlocab** [comment]
>Only difference being Catalina and not BS

This is a pretty important difference -- that's why https://github.com/Homebrew/brew/issues/10127 is relevant for the OP but not for you. i.e. it has everything to do with the original post, but not for you, so yours is actually a different issue entirely.

See also https://github.com/Homebrew/homebrew-core/issues/72448#issuecomment-790254171.

**Comment by carlocab** [comment]
I guess, to put this differently, the cause of the problem reported in the first post is that `ldns` was installed from a Catalina bottle, which isn't compatible with `openssh` installed from a Big Sur bottle.

The solution to that is to make sure we track bottles that need to be reinstalled upon OS upgrades. This is tracked in https://github.com/Homebrew/brew/issues/10127.

Your issue, on the other hand, seems to be caused by something else entirely. We'd have been able to figure that out much earlier had you filed a proper bug report, rather than piggybacked on a similar-looking but actually completely unrelated bug report. Not sure why you keep insisting your issue is the same when it quite clearly has an entirely different cause (which we can't ascertain because of the absence of a proper bug report).

**Comment by metayan** [comment]
Thanks for the clarification.

Since this issue popped up when searching for how to resolve my issue, and looked identical to what I saw with the exception of being on BS, I didn't want to file a duplicate issue. Also, the solution presented here solved my issue, which is why I raised my concern that something should be done for Catalina as well – when I was notified that this issue was being closed.

Also, being a Homebrew user and not familiar with the inner workings and intricate details – just trying to help – I find your last paragraph (after the first sentence) unnecessarily accusational and plain rude.

**Comment by Bo98** [comment]
I think it would be nice to know _why_ the ABI version is different on Catalina and Big Sur. This is not normal behaviour.

If someone figures out why, then we could adjust the formula to make it more consistent.

**Comment by carlocab** [comment]
>I find your last paragraph (after the first sentence) unnecessarily accusational and plain rude.

That's interesting -- I found your previous continued insistence that your issue is the same as the OP, and the absence of proper debugging information, also plain rude. Guess there's a lot of that going around.

>I think it would be nice to know why the ABI version is different on Catalina and Big Sur. This is not normal behaviour.

It is very strange. I tried rebuilding it on Catalina with the latest CLT, and it still shows ABI version 0, so I don't think it's related to anything specific about the macOS version. If it means anything, building it on Linux also shows an ABI version of 3.
